src/cargo/core/registry.rs

```diff
@@ -335,35 +357,58 @@ http://doc.crates.io/specifying-dependencies.html#overriding-dependencies
 }
 
 impl<'cfg> Registry for PackageRegistry<'cfg> {
-    fn query(&mut self, dep: &Dependency) -> CargoResult<Vec<Summary>> {
+    fn query(&mut self,
+             dep: &Dependency,
+             f: &mut FnMut(Summary)) -> CargoResult<()> {
         // Ensure the requested source_id is loaded
         self.ensure_loaded(dep.source_id(), Kind::Normal).chain_err(|| {
             format!("failed to load source for a dependency \
                      on `{}`", dep.name())
         })?;
 
+        // Look for an override and get ready to query the real source
         let override_summary = self.query_overrides(&dep)?;
-        let real_summaries = match self.sources.get_mut(dep.source_id()) {
-            Some(src) => Some(src.query(&dep)?),
-            None => None,
+        let mut sources = self.sources.borrow_mut();
+        let source = match sources.get_mut(dep.source_id()) {
+            Some(src) => src,
+            None => {
+                if override_summary.is_some() {
+                    bail!("override found but no real ones");
+                }
+                return Ok(())
+            }
         };
 
-        let ret = match (override_summary, real_summaries) {
-            (Some(candidate), Some(summaries)) => {
-                if summaries.len() != 1 {
-                    bail!("found an override with a non-locked list");
-                }
-                self.warn_bad_override(&candidate, &summaries[0])?;
-                vec![candidate]
+        // Query the real source, keeping track of some extra info. If we've got
+        // an override we don't actually ship up summaries. If we do ship up
+        // summaries though we be sure to postprocess them to a locked version
+        // to assist with lockfiles and conservative updates.
+        let mut n = 0;
+        let mut to_warn = None;
+        source.query(dep, &mut |summary| {
+            n += 1;
+            if override_summary.is_none() {
+                f(self.lock(summary))
+            } else {
+                to_warn = Some(summary);
             }
-            (Some(_), None) => bail!("override found but no real ones"),
-            (None, Some(summaries)) => summaries,
-            (None, None) => Vec::new(),
+        })?;
+
+        // After all that's said and done we need to check the override summary
+        // itself. If present we'll do some more validation and yield it up,
+        // otherwise we're done.
+        let override_summary = match override_summary {
```

Comment:
It's a bit hard to match this `match` with an `if` inside the lambda inside the query. Perhaps something like

```
match override_summary {
  None => { source.query(dep, &mut|summary| f(self.lock(summary)))?; Ok(()) }
  Some(overrie_summary) => {
    let mut n = 0;
    let mut to_warn = None;
    source.query(dep, &mut |summary| { n += 1; to_warn = Some(summary)})
  }
}
```

would read better? That is, pull `query` inside both arms of a `match`. Perhaps in the second arm we could even use `query_vec`, because we are going to print a warning anyway?